lib/internal/errors.js

```diff
@@ -954,6 +956,20 @@ E('ERR_TRANSFORM_ALREADY_TRANSFORMING',
 E('ERR_TRANSFORM_WITH_LENGTH_0',
   'Calling transform done when writableState.length != 0', Error);
 E('ERR_TTY_INIT_FAILED', 'TTY initialization failed', SystemError);
+E('ERR_TYPE_MISMATCH', (filename, ext, typeFlag, conflict) => {
+  const typeString =
+    typeFlag === 'module' ? '--type=module or -m' : '--type=commonjs';
+  // --type mismatches file extension
+  if (conflict === 'extension')
+    return `Extension ${ext} is not supported for ` +
+      `${typeString} loading ${filename}`;
+  // --type mismatches package.json "type"
+  else if (conflict === 'scope')
+    return `Cannot use ${typeString} because nearest parent package.json ` +
+      ((typeFlag === 'module') ?
+        'includes "type": "commonjs"' : 'includes "type": "module",') +
+      ` which controls the type to use for ${filename}`;
```

Comment:
This seems to miss an `else` case in case `conflicts` is neither `'extension'` nor `'scope'`.

Reply:
There is no usage of this ERROR where that would be the case. Is there a best practice to ensure usage of an error in a certain way?

Reply:
An error error is a new one...!

Reply:
I guess the concern here is just if we allow invalid error argument states, which I guess we always do if we don't validate error function arguments themselves.

This seems fine to me though.

Reply:
We could use a `CHECK(false)` internal macro – assuming the support for it added back in 3ed363cb36a9fb686956c0b8b2953ff08a6f0ee8 still works (cc @devsnek). Or we could use an `assert.fail`. I don't want errors to go undetected though (and causing some odd issue down the line).

Reply:
As far as I'm aware, the check macros still work. `CHECK` will abort the process, while `assert` throws an error.

Reply:
Are we ok to move forward with the current implementation? If not does someone want to push a patch to fix this?

Reply:
You can use a switch case that does `assert.fail` if the type flag does not match any known values, but I don't think it matters a lot to validate those in `internal/errors.js` unless we bring the internal error tests back.

Reply:
`assert.fail` would be totally fine. I'll push a fix. We already have a couple of assertions in some errors and there are more implicit argument checks that already caught a couple of implementation issues (we always validate the argument length passed to our errors).